### ISSUE TITLE
[BUG] Tool crashes midgame in game

### ISSUE THREAD

**Issue by benholness** [RCVolus/league-observer-tool#21]
**Describe the bug**
The tool crashes during in game randomly without any error message popup or indication in the League Prod Toolkit console.

**To Reproduce**
1. Load up League of Legends and start up the League Prod Toolkit via start.bat
2. Once server is started, load up League Observer Tool and sync up all elements for both Client and In Game tabs.
3. Start up game (happens both for live spectator and replays).
4. Without any indiciation, the tool crashes midgame (no specific time it happens).

**Expected behavior**
To not crash at all or at the very least show the error message of what the issue at hand is/why it crashed.

**System:**
 - OS: Windows 10
 - Version: 10.0.19045
 - League Observer Tool Version: 6.2.5

**Additional context**
Main Log (the error message in here could be related to the bug I've described above, but I'm not 100% sure as it happened on 06/08/2023 as well and I don't see any errors logged even though it crashed that day too): 
[main.log](https://github.com/RCVolus/league-observer-tool/files/12273013/main.log)

**Comment by benholness** [comment]
As an extra bit of info to see this live in action, this happened when I was casting some games on stream multiple times (you can see it around the 40min 40 second mark where the timer freezes at 07:02 and it stops updating the scoreboard). The program just closes itself randomly for some reason.

https://www.twitch.tv/videos/1889182971

**Comment by sdomi4** [comment]
Is there no output on the console before the crash or does the console just close before you can see / copy out any errors?

**Comment by benholness** [comment]
So there are no errors in the console for League Prod Toolkit and the League Observer Tool just closes itself.

**Comment by benholness** [comment]
I'm casting another game tonight so I'll see if anything happens but this has happened ever since I updated the League Observer Tool to 6.2.5

**Comment by sdomi4** [comment]
Are you positive this only started happening with 6.2.5? That would allow us to narrow down the issue.

**Comment by benholness** [comment]
I'm 99% sure I didn't have the program crash like this before the update. Most of the errors I've had in the past were Farsight related and showed up in the League Pro Toolkit console.

**Comment by benholness** [comment]
Hey @sdomi4 and @Himyu, done a cast yesterday after I updated League Prod Toolkit to 1.6.1 and it didn't crash once.

I assume 6.2.5 didn't like 1.6.0 of League Prod Toolkit 👍

**Comment by Himyu** [comment]
Ok perfect, 
I will have a look into it but if that fixed your issue that's great.

**Comment by MDvorky** [comment]
It keeps crashing and freezing the scoreboard. Also shrinking my middle logo of the series. Can you please help me somehow? I did not have this issue two verzii s before so i dont know what to do.

**Comment by MDvorky** [comment]
Also recieving this message - don't know what to do, because i did update the config and live.events in install folder.  There was additional dragonsoulgiven line. Don't know if that's right or not.
![image](https://github.com/RCVolus/league-observer-tool/assets/104410235/97c19614-7df5-40b4-a550-67857be258ca)
And as well this is kinda strange - is this normal?
![image](https://github.com/RCVolus/league-observer-tool/assets/104410235/bf0bb37c-8eb3-4b4f-a34f-b840c54c2c37)

**Comment by sdomi4** [comment]
Please keep things to one place, we are looking into adding logs to troubleshoot the issue further.

**Comment by MDvorky** [comment]
Okay, but i don't know what can i do right now. Just wait - tried to do some thing, i will test it today and let you know.

**Comment by MDvorky** [comment]
It's still crashing. I dont know what else i Can do.

**Comment by Himyu** [comment]
As we told you before we can not help you if you do not provide details of the problems. We need more descriptions, logs, etc to find out what is wrong if you don't offer that we can not help you.

**Comment by MDvorky** [comment]
If your can tell me where Can i find logs or whatever, i will show it to you. But from where i Can pull logs if app crashes?

**Comment by Himyu** [comment]
you can find your logs under 'AppData\Roaming\league-observer-tool\logs'

**Comment by sdomi4** [comment]
Is the error logged at `2023-08-15 12:24:23.945` the same time when the observer tool crashed? Or was the crash at `2023-08-15 12:42:03.429`?

**Comment by MDvorky** [comment]
crash happened at 12:42:03 but i started program at 12:24.

**Comment by MDvorky** [comment]
I found this in Config - maybe this was causing the issue:

{
    "server-ip": "127.0.0.1",
    "server-port": 3003,
    "server-api-key": "RCVPT-YV7BQ3X-AAX4AQ2-G17RDCA-EP3M4ME",
    "live-events-port": 34243,
    "replay-send-information": true,
    "league-install-path": "C:\Riot Games\League of Legends"
}

**Comment by MDvorky** [comment]
so i replaced it and put the thingies which are in the guide.

**Comment by MordredXYZ** [comment]
I'd also like to report I'm getting CONSISTENT crashes from the observer tool, and have been for quite some time but up until I saw this thread, believed it to be my machines.

I've had the tool crash twice while testing my setup. Scratch that, it just crashed as I was typing that, and with ZERO log output.

Just crashed now at 8:24PM, and last log entry was when I reopened it, AGAIN, at 8:15pm. [main.log](https://github.com/RCVolus/league-observer-tool/files/12395053/main.log)

With multiple crashes a game, this is... unreliable at best.

**Comment by MordredXYZ** [comment]
The tool also consistently prompts me to fix my game config, even though nothing has changed and my client is set up correctly. That is likely another, separate issue however.

**Comment by Himyu** [comment]
Just a couple of questions to narrow down the origin of the bug.
The prod tool itself is no issue and does not show any errors?

Are you using the auth feature or not?

Are you using all features of the observer tool?

Can you provide your config without sensitive data to check so settings?

**Comment by Himyu** [comment]
> The tool also consistently prompts me to fix my game config, even though nothing has changed and my client is set up correctly. That is likely another, separate issue, however.

Yes that would be a separate issue please.

**Comment by MDvorky** [comment]
@Himyu If i may, im recieving some error in prod tool. 
I'm using auth
I'm using all the features - whic i can. And this is my config.
I already past the log file here but i can do i
t again. Today it crashes again for a few times. I was forced to use not my great overlay but basic which was pitty.

**Comment by sdomi4** [comment]
403 usually indicates an issue with the API key. There are also some issues with the APIs on Riots side (spectator-v4 which provides info on runes and some other things is broken)

**Comment by MDvorky** [comment]
But I did not have it few versions before. So there is nothing I can do about it?

**Comment by benholness** [comment]
> 

I'm getting this same error btw when I load up the program. I checked my LiveEvents.ini and it has this. Is this all the correct elements the file needs to have?

```
OnMinionKill
OnNeutralMinionKill
OnDamageGiven
OnKillDragon_Spectator
OnKillRiftHerald_Spectator
OnKillWorm_Spectator
OnTurretPlateDestroyed
OnDragonSoulGiven
```

**Comment by benholness** [comment]
Every time I load the program it prompts me to say my game configs are missing or not complete. And when I click yes it just adds:

```
OnKillDragon_Spectator
OnKillRiftHerald_Spectator
OnKillWorm_Spectator
OnTurretPlateDestroyed
OnDragonSoulGiven
```

to the LiveEvents.ini but doesn't solve the error. Any thoughts on this? Let me know if there is any additional information you guys need and I'll be happy to pass it on to help resolve this.

<img width="279" alt="CaptureError" src="https://github.com/RCVolus/league-observer-tool/assets/28854222/68c0ea3a-c8d6-4a77-9be7-4a8660e11ac9">

**Comment by benholness** [comment]
In addition to this the overlays work as expected (the scoreboard is fine) however level up popups, item popups and the ingame leaderboards don't work (leaderboards are empty).

**Comment by Himyu** [comment]
regarding the 403 issue, there is nothing we can do since that means either the API key is not valid or an action is blocked by Riot. we can check out what exactly is causing this, but we probably cant do anything about that.

**Comment by benholness** [comment]
No worries @Himyu, thank you for the update. We appreciate everything you guys are doing for this <3

**Comment by benholness** [comment]
Thank you, let me know if there's anything I can do to help provide you with the needed info to replicate it.

**Comment by Himyu** [comment]
is your game.cfg locked for modification? A lot of people do that so the riot client does not mess up their configs.
and is that happening on the normal server or tournament realm?

**Comment by benholness** [comment]
Hey @Himyu, so I've had a look and I can edit my game.cfg file but not 100% sure how to check if it's locked. Is there a way to check?

And this is happening in normal server games (i.e custom tournament draft games and spectating normal games)

**Comment by MDvorky** [comment]
Me as well have this issue in normal games.

**Comment by Himyu** [comment]
Okay, it will take some time to find and fix this issue. We will update you as soon as we know more about it.

**Comment by MDvorky** [comment]
So yesterday i downloaded new version of both - but tool crashed me at 2. minute. If i can send the logs, i will but can you help? It is pitty to not having this app running. :/

**Comment by sdomi4** [comment]
We are working on narrowing down the issue – unfortunately the logs don't catch the issue at the moment.
You can try disabling the LiveEvents connector for better stability of the observer tool, but will lack Objective Kills and Platings showing up properly.

**Comment by MDvorky** [comment]
I'm not using Platings anyway - i'm focusing more on scoreboard and baron timers :) Thank you for amazing job.

**Comment by yourideraadt** [comment]
Hey, so I indeed also encountered the error, and was able to replicate it somewhat easily. By watching back a previous game in spectator, and skipping back and forth constantly through the match it would crash relatively quickly (within 10 seconds). I hope this helps to find what is causing the issue.

**Comment by sdomi4** [comment]
Skipping around in time is not really an intended use case for the observer tool, as it's expected to run on the main obs client; so that's not something we account/test for.

**Comment by FireFoXz09** [comment]
Just to update or add to maybe help track. Fresh update on my side of the observer tool and the prod toolkit on a new PC. First time spectating a game, the observer tool crash at around 38 minutes mark. I'll do more test with more games but no log available about this crash. I will try disabling the LiveEvents connector to see if this add some stability or not.

**Comment by yourideraadt** [comment]
From my experience it has been specifically related to farsight. I've been extensively testing and using the program while developing a custom overlay, and crashes only occured when farsight was enabled. So I can 99% confidently say it is related to that. And this was not with skipping back and forth as mentioned in my previous comment.

**Comment by sdomi4** [comment]
We haven't been able to fully narrow down the issue, but version 6.5.2 should hopefully be more resilient to crashes (and log more information surrounding crashes).
If you continue seeing crashes (we haven't been able to consistently reproduce any without skipping around the timeline) please add a comment with the log file, and note whether or not you were skipping around the timeline.

**Comment by FireFoXz09** [comment]
Still crashes randomly with patch 6.5.2 (on this example, at 10 minutes 3 seconds IGT). Here's the log. It was from a downloaded replay. I did not touch the timeline, or skip time, or go faster, etc. I just let the replay go and wait until the Observer Tool stop running and close. The toolkit still run and don't say any error.

[main.log](https://github.com/RCVolus/league-observer-tool/files/12851245/main.log)

**Comment by yourideraadt** [comment]
I can definitely confirm that it is related to farsight, as it still happens in the latest version. I only had the farsight module synced, and none of the others, and after just spectating a game for a couple of minutes it spontaneously crashed.
I suspect maybe it could be windows version related, I am using windows 10 as well as the initial poster, but otherwise I have no clue, as the crashing is highly inconsistent.

**Comment by MDvorky** [comment]
I can add as well log, if you can tell me where can i found it. Today i had around 34 minutes and then probably after baron it crashes and i had to restart it.

**Comment by JustSxm** [comment]
Hello so I come here to help debug the issue

No interaction in spectator mode causes this issue. I tried to touch nothing while the game was going on and the observer crashed, I looked into the log and nothing out of the ordinary was logged, no errors.

The config issue happens when your LiveEvents.ini contains more events than the observers tool says to have, if you remove the ones not in the github "tutorial" you will not receive the error


Im not sure what could be causing the issue, but I can tell you that the observer tool crashed after the nexus exploded and the players were disconnecting, I dont think a live event is the cause of this issue.

It also crashed at random times, but for now the logs the last thing in them was

[2023-10-30 11:46:46.184] [info]  (in-game-live-events) [
  {
    eventname: 'OnTurretPlateDestroyed',
    other: 'Turret_T2_L_03_A',
    otherTeam: 'Chaos',
    source: 'Minion_T100L2S23N0064',
    sourceTeam: 'Order'
  }
]

**Comment by sdomi4** [comment]
Yes, the issue is almost certainly related to farsight, not LiveEvents (we assumed LiveEvents was to blame since it seemed like a game state issue in combination with no error handling being implemented for LiveEvents).

**Comment by JustSxm** [comment]
The function who causes the crash is ``makeSnapshot`` in index.cc. Ill keep investigating

**Comment by JustSxm** [comment]
So from what I found in ``CreateSnapshot`` theres a call to ``ReadObjects`` who in a for-loop (line 180 in Farsight.cc) Tries to load something from memory and that fails. Someone who understood C++ better than me and took time to understand what its trying to do and why whatever it is trying to load fails could fix it, for now I will attempt to just do a bandage that if the snapshot fails, it tries again. If that does not work, atleast I will have debugged the issue for anyone interested in fixing it.

**Comment by JustSxm** [comment]
Unfortunately even when trying to apply a bandage fix it does not work.

When trying to read something in the memory it fails. (what I said above) For a reason I dont know, this after breaks everything. Retrying to make a snapshot will never work after this occurs. 

The only solution will be that someone makes a proper fix

**Comment by JustSxm** [comment]
Hello, so the error kept haunting me and I went back at it today and found the reason AND a solution for the bug.

Every second LoL observer tool is trying to make a snapshot, this snapshot reads timers,etc. Though, we are trying to do a "deepload" snapshot everytime. A deepload snapshot includes the health, the mana, and the display name of the users. Im unsure why we need to it every second, as I feel like those are not needed every second.  (I think we only need to do it once to gather the usernames for Gold and XP Level of prod-toolkit)

Anyway, the crash happens because gathering the usernames from memory does not always work, when this fails the application crash. Putting a try/catch around it solves the issues.

I will just make sure of it by watching 2 whole game without a crash, if this occurs I will create a pull request

**Comment by sdomi4** [comment]
Is the crash actually caused by the [memory read](https://github.com/floh22/native-farsight-module/blob/969b36ff49828fe6e6b209efb8040725db4174f8/src/unit/GameObject.cc#L65) when `deepLoad` is set to true? As in, which code portion did you enclose in a try/catch block to solve the issue?

**Comment by JustSxm** [comment]
> Is the crash actually caused by the [memory read](https://github.com/floh22/native-farsight-module/blob/969b36ff49828fe6e6b209efb8040725db4174f8/src/unit/GameObject.cc#L65) when `deepLoad` is set to true? As in, which code portion did you enclose in a try/catch block to solve the issue?

Yes, this portion of code is the reason of crashes, I just went through a full game without a crash (except I just had some kind of disconnect from my terminal to Lol Observer, so Ill need to test out how it works out when installed)

`int displayNameLength = Memory::ReadDWORD64FromBuffer(buff, Offsets::ObjDisplayNameLength);

            if (displayNameLength < 16)
            {
                displayName.resize(displayNameLength);
                memcpy(displayName.data(), &buff[Offsets::ObjDisplayName], displayNameLength);
            }
            else
            {
                char displayNameBuff[50];
                Memory::Read(hProcess, Memory::ReadDWORD64FromBuffer(buff, Offsets::ObjDisplayName), displayNameBuff, 50);
                displayName.resize(displayNameLength);
                memcpy(displayName.data(), &displayNameBuff[0], displayNameLength);
            }`

**Comment by JustSxm** [comment]
Sorry I didnt read your question correctly, No it isn't only that memory reading, both of them in my preview can fail.

The memcpy can also fail in the if case

**Comment by sdomi4** [comment]
Good news everyone – we are confident the crashes are now fixed.
Thanks to @JustSxm for helping figure out where the problem was coming from and Lörs for fixing the issue. Special shoutout to Querijn for jumping in and providing insights as well.

The updated observer tool with the fixed version of farsight can be [downloaded here](https://github.com/RCVolus/league-observer-tool-codebase/releases/tag/v6.7.1), **a manual install is needed.**
Note this is not the usual place for the releases, we are in the middle of merging this repository with the codebase repository. Due to this, the autoupdate will not work to get the newest version at the moment, but should take care of it soon.

**Comment by MDvorky** [comment]
You are legends! <3

**Comment by FireFoXz09** [comment]
Tested it in a 45 minutes + game. No crash, all data is correctly displayed. I even moved in the game many times (jumping 10+ minutes back and forth to test baron power play and pop-ups) for 2 hours long non-stop in the same game without any issue.
